opentelemetry-sdk/src/trace/sampler.rs

```diff
@@ -172,6 +172,25 @@ impl ShouldSample for Sampler {
         attributes: &[KeyValue],
         links: &[Link],
     ) -> SamplingResult {
+        // Checking suppression mode in sampler is more efficient than checking
+        // it in the span processor as this allows earlier short-circuiting.
+        // Also, it is currently *NOT* possible to check suppression mode or
+        // even access current context in the span processor's OnEnd method. See
+        // https://github.com/open-telemetry/opentelemetry-rust/issues/2871
+        // TODO: Can we move this even earlier, i.e. in the Tracer itself?
+        // TODO: Check and fix: why this methods gets Option<&Context> and not
+        // Context?
+        if let Some(parent_context) = parent_context {
+            if parent_context.is_telemetry_suppressed() {
```

Comment:
I'm not so familiar with the trace SDK implementation but it looks like we [don't always](https://github.com/open-telemetry/opentelemetry-rust/blob/297146701d483397432323c6e6cdef86e9da9999/opentelemetry-sdk/src/trace/tracer.rs#L216-L229) call `should_sample`. I don't know if that `tracing-opentelemetry` pre-sampling scenario need the suppression check.

Reply:
Good point. Those should be fixed/cleaned up. I also put a TODO to move this check even earlier than sampling.